### PR TITLE
expression: ECB/CBC modes with 128/192/256-bit key length for AES

### PR DESCRIPTION
**Pull request by mccxj**
### What problem does this PR solve? 

Fix https://github.com/pingcap/tidb/issues/7300. Support the `init_vector` argument for builtin function `AES_ENCRYPT`/`AES_DECRYPT`

### What is changed and how it works?

- add system variable 'block_encryption_mode' support

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change

### CODE REVIEW

File: expression/builtin_encryption.go [hunk 1]
```diff
@@ -112,9 +130,42 @@ func (b *builtinAesDecryptSig) evalString(row chunk.Row) (string, bool, error) {
...
+	}
+	if mode.ivRequired {
+		if len(b.args) != 3 {
+			return "", true, ErrIncorrectParameterCount.GenByArgs("aes_decrypt")
+		}
+		iv, isNull, err = b.args[2].EvalString(b.ctx, row)
+		if isNull || err != nil {
+			return "", true, errors.Trace(err)
+		}
+		if len(iv) < aes.BlockSize {
```
Comment (lysu) [inline expression/builtin_encryption.go]: what about give an `IVSize` to help understand~? although our `IVSize = BlockSize = 16`

File: expression/builtin_encryption.go [hunk 2]
```diff
@@ -112,9 +130,42 @@ func (b *builtinAesDecryptSig) evalString(row chunk.Row) (string, bool, error) {
...
+		}
+	}
+
+	key := encrypt.DeriveKeyMySQL([]byte(keyStr), mode.keySize)
+	var plainText []byte
+	switch mode.modeName {
+	case "ecb":
+		plainText, err = encrypt.AESDecryptWithECB([]byte(cryptStr), key)
+	case "cbc":
+		plainText, err = encrypt.AESDecryptWithCBC([]byte(cryptStr), key, []byte(iv))
```
Comment (lysu) [inline expression/builtin_encryption.go]: It seems some duplicate logic in `AESDecryptWithECB` and `AESDecryptWithCBC`, `AESEncryptWithECB` and `AESEncryptWithCBC`~

what about use `AESDecrypt`, `AESEncrypt`, then inversion of control, pass a `cipher.BlockMode` argument?

File: expression/builtin_encryption.go [hunk 3]
```diff
@@ -26,13 +26,16 @@ import (
...
+	"crypto/aes"
 	"github.com/juju/errors"
 	"github.com/pingcap/tidb/mysql"
 	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/util/auth"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/encrypt"
+	"strings"
```
Comment (lysu) [inline expression/builtin_encryption.go]: please move `strings` to line 28(it's TiDB rule, standard package need in first import group ^ ^)

File: expression/builtin_encryption.go (hunk 2 above)
Comment (mccxj) [reply expression/builtin_encryption.go]: I add "type blockModeBuild func(block cipher.Block) cipher.BlockMode" to create BlockMode argument.

File: expression/builtin_encryption.go (hunk 1 above)
Comment (mccxj) [reply expression/builtin_encryption.go]: OK. add const IVSize = aes.BlockSize.

**Comment by mccxj** [comment]
@lysu PTAL.

File: expression/builtin_encryption.go [hunk 4]
```diff
@@ -112,9 +133,42 @@ func (b *builtinAesDecryptSig) evalString(row chunk.Row) (string, bool, error) {
...
-	plainText, err := encrypt.AESDecryptWithECB([]byte(cryptStr), key)
+	var iv string
+	modeName, _ := b.ctx.GetSessionVars().GetSystemVar(variable.BlockEncryptionMode)
+	mode, exists := aesModes[strings.ToLower(modeName)]
+	if !exists {
+		return "", true, errors.Errorf("unsupported block encryption mode - %v", modeName)
+	}
+	if mode.ivRequired {
+		if len(b.args) != 3 {
+			return "", true, ErrIncorrectParameterCount.GenByArgs("aes_decrypt")
```
Comment (zz-jason) [inline expression/builtin_encryption.go]: This check should be done in `aesDecryptFunctionClass.getFunction()`

File: expression/builtin_encryption.go [hunk 5]
```diff
@@ -112,9 +133,42 @@ func (b *builtinAesDecryptSig) evalString(row chunk.Row) (string, bool, error) {
...
+		}
+		iv, isNull, err = b.args[2].EvalString(b.ctx, row)
+		if isNull || err != nil {
+			return "", true, errors.Trace(err)
+		}
+		if len(iv) < IVSize {
+			return "", true, errIncorrectArgs.Gen("The initialization vector supplied to aes_decrypt is too short. Must be at least 16 bytes long")
```
Comment (zz-jason) [inline expression/builtin_encryption.go]: Is this error message the same with MySQL?

Comment (lysu) [reply expression/builtin_encryption.go]: Yes..I had checked, messages is same to mysql.

but in strict, errorcode returned by this PR has some different to mysql, expect this point, new code using `errors.Errorf` maybe should add a new `ErrXXX.GenByArgs`

File: expression/builtin_encryption.go [hunk 6]
```diff
@@ -68,10 +71,24 @@ var (
...
+const IVSize = aes.BlockSize
+
+type aesModeAttr struct {
+	modeName   string
+	keySize    int
+	ivRequired bool
+}
+
+var aesModes = map[string]*aesModeAttr{
+	//TODO support more modes
```
[the remark this answers is not in the document]
Comment (mccxj) [reply expression/builtin_encryption.go]: OK. see https://github.com/pingcap/tidb/issues/7490

File: expression/builtin_encryption.go [hunk 7]
```diff
@@ -112,9 +154,64 @@ func (b *builtinAesDecryptSig) evalString(row chunk.Row) (string, bool, error) {
...
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
+	}
+
+	iv, isNull, err := b.args[2].EvalString(b.ctx, row)
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
+	}
+	if len(iv) < aes.BlockSize {
+		return "", true, errIncorrectArgs.Gen("The initialization vector supplied to aes_decrypt is too short. Must be at least 16 bytes long")
```
Comment (zz-jason) [inline expression/builtin_encryption.go]: s/16/aes.BlockSize/?

File: expression/builtin_encryption.go [hunk 8]
```diff
@@ -81,20 +103,40 @@ func (c *aesDecryptFunctionClass) getFunction(ctx sessionctx.Context, args []Exp
...
+		return nil, errors.Errorf("unsupported block encryption mode - %v", blockMode)
+	}
+	if mode.ivRequired {
+		if len(args) != 3 {
+			return nil, ErrIncorrectParameterCount.GenByArgs("aes_decrypt")
+		}
+		return &builtinAesDecryptIVSig{bf, mode}, nil
+	} else if len(args) == 3 {
+		// For modes that do not require init_vector, it is ignored and a warning is generated if it is specified.
+		ctx.GetSessionVars().StmtCtx.AppendWarning(errWarnOptionIgnored.GenByArgs("IV"))
```
Comment (zz-jason) [inline expression/builtin_encryption.go]: It seems this warning should be appended in the function execution phase for every input row:
```sql
MySQL(localhost:3306) > select aes_encrypt(a, b, c) from t;
+----------------------+
| aes_encrypt(a, b, c) |
+----------------------+
| 0SF&Qٜ               |
| 0SF&Qٜ               |
| 0SF&Qٜ               |
| 0SF&Qٜ               |
+----------------------+
4 rows in set, 4 warnings (0.00 sec)

MySQL(localhost:3306) > show warnings;
+---------+------+---------------------+
| Level   | Code | Message             |
+---------+------+---------------------+
| Warning | 1618 | <IV> option ignored |
| Warning | 1618 | <IV> option ignored |
| Warning | 1618 | <IV> option ignored |
| Warning | 1618 | <IV> option ignored |
+---------+------+---------------------+
4 rows in set (0.00 sec)
```

File: expression/builtin_encryption.go (hunk 7 above)
Comment (mccxj) [reply expression/builtin_encryption.go]: done.

File: expression/builtin_encryption.go (hunk 8 above)
Comment (mccxj) [reply expression/builtin_encryption.go]: moved to execution phase, and new testcase is added.

**Comment by mccxj** [comment]
@zz-jason PTAL.

File: expression/builtin_encryption.go [hunk 9]
```diff
@@ -25,10 +25,13 @@ import (
 	"fmt"
 	"hash"
 	"io"
+	"strings"
 
+	"crypto/aes"
```
Comment (zz-jason) [inline expression/builtin_encryption.go]: please put the standard packages together and add an empty line between the third-part packages.

File: util/encrypt/aes.go [hunk 10]
```diff
@@ -120,26 +122,49 @@ func PKCS7Unpad(data []byte, blockSize int) ([]byte, error) {
...
+	return aesDecrypt(cryptStr, key, func(block cipher.Block) cipher.BlockMode {
+		return newECBDecrypter(block)
+	})
+}
+
+// DeriveKeyMySQL derives the encryption key from a password in MySQL algorithm.
+// See https://security.stackexchange.com/questions/4863/mysql-aes-encrypt-key-length.
+func DeriveKeyMySQL(key []byte, blockSize int) []byte {
+	rKey := make([]byte, blockSize)
+	rIdx := 0
```
Comment (zz-jason) [inline util/encrypt/aes.go]: why prefix `Idx` and `Key` with `r`?

File: util/encrypt/aes.go [hunk 11]
```diff
@@ -120,26 +122,49 @@ func PKCS7Unpad(data []byte, blockSize int) ([]byte, error) {
...
+
+// AESDecryptWithCBC decrypts data using AES with CBC mode.
+func AESDecryptWithCBC(cryptStr, key []byte, iv []byte) ([]byte, error) {
+	return aesDecrypt(cryptStr, key, func(block cipher.Block) cipher.BlockMode {
+		return cipher.NewCBCDecrypter(block, iv)
+	})
+}
+
+// aesDecrypt decrypts data using AES.
+func aesDecrypt(cryptStr, key []byte, build blockModeBuild) ([]byte, error) {
```
Comment (zz-jason) [inline util/encrypt/aes.go]: It's better to pass `cipher.BlockMode` instead of `blockModeBuild`

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

File: util/encrypt/aes.go (hunk 10 above)
Comment (mccxj) [reply util/encrypt/aes.go]: I have not changed it. It seems that code position change affects 'git diff'.see line#163.

**Comment by jackysp** [comment]
@mccxj , I've added a word "Fix" before the issue link in the description of this PR, then the issue could be closed automatically when this PR is merged.

File: expression/builtin_encryption.go (hunk 9 above)
Comment (mccxj) [reply expression/builtin_encryption.go]: done.

File: util/encrypt/aes.go (hunk 11 above)
Comment (mccxj) [reply util/encrypt/aes.go]: AESEncryptWithCBC/AESDecryptWithCBC is similar with AESEncryptWithECB/AESDecryptWithECB with different cipher.BlockMode, CBC needs `iv` but ECB does not, so blockModeBuild is the hook method to build cipher.BlockMode.

Comment (zz-jason) [reply util/encrypt/aes.go]: IMHO, if we can remove this hook, pass the required `cipher.BlockMode` directly, the code would be easier to read and maintain.

**Comment by zz-jason** [comment]
@mccxj CI failed.

File: expression/builtin_encryption.go [hunk 12]
```diff
@@ -68,10 +73,27 @@ var (
 	_ builtinFunc = &builtinUncompressedLengthSig{}
 )
 
-// TODO: support other mode
-const (
-	aes128ecbBlobkSize = 16
-)
+// ivSize indicates the initialization vector supplied to aes_decrypt
+const ivSize = aes.BlockSize
```
Comment (lamxTyler) [inline expression/builtin_encryption.go]: Not used?

**Comment by mccxj** [comment]
@zz-jason @lamxTyler @jackysp PTAL

**Pull request merged**